content/2019/vim-and-the-shell/index.md

```diff
@@ -0,0 +1,227 @@
+---
+
+> Don't love me, kid. Love your mom. I just play football.
+>
+> – Juan Roman Riquelme
+
+There is a day in the life of a man when he realizes that the traditional way of using a computer is flawed, that graphical interfaces were a bad idea, that the mouse is an absurd invention, that a life away from the home row is not a life worth living.
+
+The UNIX philosophy becomes a religion when you are hit with the realization that best solutions are not provided by gigantic almighty monoliths, but are the joint effort of several minimal, highly specialized components. Each solves one part of the problem, just one, but solves it well.
+
+In this post I'll explore the idea that vim is just another member of the GNU/Linux constellation of tools, perfectly integrated in its environment and the underlying shell. Vim's built-in interaction capabilities with external programs allow for very interesting synergies. Interplays that become particularly powerful in conjunction with advanced text processing utilities and filters
+
+# Command Line Text Processing
+
+```text
+$ cat fruits.txt
+oranges 5
+apples 7
+blueberries 15
+bananas 4
+ananas 6
+```
+
+There are endless possibilities when it comes to fruit lists. For example, you can format them as a table:
+
+```text
+$ cat fruits.txt | column -t
+oranges      5
+apples       7
+blueberries  15
+bananas      4
+ananas       6
+```
+
+sort them by piece number:
+
+```text
+$ cat fruits.txt | column -t | sort -nrk 2
+blueberries  15
+apples       7
+ananas       6
+oranges      5
+bananas      4
+```
+
+or, print how many fruits there are in total:
+
+```text
+$ cat fruits.txt | awk '{sum += $2} END { print sum }'
+37
+```
+
+You can filter and transform text in virtually any way using the command line, you can remove duplicate lines with the absolute poetry of a command `awk '!visited[$0]++'` [(for the curious)](https://opensource.com/article/19/10/remove-duplicate-lines-files-awk), change curly braces to parenthesizes with `tr '{}' '()'`, or delete the first three characters in each line with `cut -c3-`.
+
+The point is, command line text processing is *powerful*. It's hard to describe the beauty of executing a chain of commands and watching it, pipe after pipe, transform your text like a perfect symphony of streams; the magic of doing so much work with so little keystrokes. The fact, however, that you are a vim user makes me believe that you know this feeling very well.
+
+Talking about vim, wouldn't it be awesome if you could use advanced text processing tools, like `awk`, `tr`, `jq`, `bc`, as you edit your vim buffer?
+It probably doesn't come as a surprise to you when I tell you that that in fact, you can.
+
+## The Big Bang
+
+Vim integrates smoothly with external programs, the underlying shell, and the UNIX toolchain. Specifically, you can execute commands from vim, read the standard output of a command into your buffer, and use your buffer (or parts of it) as standard input to a shell program.
```

Comment:
Here is another place where mentioning a specific technology (UNIX, here), while fitting, is too near-sighted. Vim can do all that on any system with a shell, including Windows.